### PR TITLE
fix: `RuntimeError` for UCP large DP

### PR DESCRIPTION
**Pull request by saforem2**
We encountered a strange bug when attempting to convert checkpoints (created with DP=768) to universal format.

An overview of the bug as well as a detailed description of the proposed fix is written up in:

[argonne-lcf / Megatron-DeepSpeed / ALCF / notes / `universal_checkpoint_bug.md`](https://github.com/argonne-lcf/Megatron-DeepSpeed/blob/main/ALCF/notes/universal_checkpoint_bug.md)

### CODE REVIEW

**Comment by saforem2** [comment]
@loadams thanks for the formatting fix!

Also just wanted to say no rush on this, I spoke briefly with @minjiazhang before the holidays about this issue and mentioned I would write up a more complete description of what I was seeing.

Maybe some minor thoughts:


I think the change in deepspeed / checkpoint / `deepspeed_checkpoint.py`,
e.g. passing the `strip_tensor_paddings` argument through to the `self.zero_checkpoint.get_state_for_rank` call (shown below):

```diff
-    def get_zero_checkpoint_state(self, pp_index, tp_index, dp_index) -> dict:
+    def get_zero_checkpoint_state(self, pp_index, tp_index, dp_index, strip_tensor_paddings: bool = True):
         return self.zero_checkpoint.get_state_for_rank(pp_index=pp_index,
                                                        tp_index=tp_index,
                                                        dp_index=dp_index,
                                                        keys_to_ignore=[PARAM_SHAPES])
                                                        keys_to_ignore=[PARAM_SHAPES],
+                                                       strip_tensor_paddings=strip_tensor_paddings)
```

✅ is OK since this just passes the argument through.


However, I'm a bit less sure about this change:


```diff
 sd = ds_checkpoint.get_zero_checkpoint_state(pp_index=pp_index,
                                                  tp_index=tp_index,
                                                  dp_index=dp_index,
+                                                 strip_tensor_paddings=False)
```

since I'm not completely clear on how the internals of this
[`_strip_tensor_paddings()`](https://github.com/microsoft/DeepSpeed/blob/3573858e7ce2c723b8c43231c6c6b0cf97dca2fc/deepspeed/checkpoint/zero_checkpoint.py#L95-L112) function work.

For our purposes, setting this to False, and thereby skipping the:

```python
if strip_tensor_paddings:
    self._strip_tensor_paddings(sd)
```

block in the [`get_state_for_rank`](https://github.com/microsoft/DeepSpeed/blob/3573858e7ce2c723b8c43231c6c6b0cf97dca2fc/deepspeed/checkpoint/zero_checkpoint.py#L53-L73) call seems to work, though I'm not really sure why.

**Comment by xylian86** [comment]
@saforem2 Thank you for reporting the issues!

#### Let me first explain why we need [_strip_tensor_paddings()](https://github.com/microsoft/DeepSpeed/blob/3573858e7ce2c723b8c43231c6c6b0cf97dca2fc/deepspeed/checkpoint/zero_checkpoint.py#L95-L112) function:

In ZeRO, when optimizer states are partitioned across different ranks, padding is added to align NCCL all-gather send buffers to 4-byte boundaries, so when DP=768, the alignment becomes 4 / 2 * 768 = 1536 ([Code Reference](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/stage_1_and_2.py#L356)).

Universal Checkpointing follows a padding-free design principle. When converting to Universal Checkpoints, we must remove these paddings because:

- Different GPU cluster configurations will require different padding sizes
- New padding values should be calculated dynamically during loading
- Retaining old padding could lead to functional or correctness issues when loading checkpoints

Therefore, in this PR, skipping [_strip_tensor_paddings()](https://github.com/microsoft/DeepSpeed/blob/3573858e7ce2c723b8c43231c6c6b0cf97dca2fc/deepspeed/checkpoint/zero_checkpoint.py#L95-L112) might avoid immediate conversion issues, but I suspect it might be issues when loading universal checkpoints later if we keep the padding, either functional issue or correctness issue.

---

#### Why the conversion works for smaller DP size but failed on DP size equals to 768.

Let me explain the cause of the issue, from the [log](https://github.com/argonne-lcf/Megatron-DeepSpeed/blob/main/ALCF/notes/universal_checkpoint_bug.md) you provided (thank you for providing such detailed logs!!!), I can see the model config (number of layers = 32, hiddle size = 4096), which ia a ~7B model, and you trained with ZeRO1 with DP size 768. 

Param groups will be created based on weight decay condition (regularized vs non regularized) and learning rate scale condition (args.lr vs lr_mult * args.lr) ([Code Reference](https://github.com/argonne-lcf/Megatron-DeepSpeed/blob/main/megatron/optimizer/__init__.py#L18-L103)). 

For one group, the flatten tensor numel is 266240, the alignment is 1536, so the required padding = 1536 * 174 - 266240 = 1024, per-rank partition size is 348 (1536 * 174 / 768). The current implementation assumes that the number of elements assigned to each rank will exceed the padding size. However, in this case, we have 348 < 1024, violating this assumption.

---

The correct way is to fix this issue in [_strip_tensor_paddings()](https://github.com/microsoft/DeepSpeed/blob/3573858e7ce2c723b8c43231c6c6b0cf97dca2fc/deepspeed/checkpoint/zero_checkpoint.py#L95-L112) function to handle such corner cases.

@tjruwase @tohtana Please let me know if you have any suggestons or feedbacks on it.

**Comment by xylian86** [comment]
More specifically, the issue arises from the following lines in [ZeRO 1 Engine](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/runtime/zero/stage_1_and_2.py#L369-L373)

```
            if partition_id == dist.get_world_size(group=self.real_dp_process_group[i]) - 1:
                padding = self.bit16_groups_flat[i].numel() - orig_group_numel
            else:
                padding = 0
```

that has the assumption that only the last rank will have the padding.

**Comment by loadams** [comment]
FYI @saforem2 - since we switched from CLA to DCO you'll need to sign the commits with the steps from [here](https://github.com/deepspeedai/DeepSpeed/pull/6918/checks?check_run_id=36821629555)

**Comment by saforem2** [comment]
This appears to be fixed, checkpoints which failed to convert previously now seem to be converting successfully.

e.g.

```bash
#[🐍 aurora_nre_models_frameworks-2024.2.1_u1](👻 aurora_nre_models_frameworks-2024.2.1_u1)
#[09:17:36 AM][x4609c0s6b0n0][/f/A/f/p/d/Megatron-DeepSpeed][🌱 main][✓] [⏱️ 1m24s]
$ ckpt_dir=checkpoints/ws768_ds_stage1_nl32_hs4096_mb4_seq4096_gb3072_sp1_pp1_tp1_bf16_optadamw_lr0.00020_lwf0.05/global_step95800 && gs=$(cat "${ckpt_dir}/latest_checkpointed_iteration.txt") && echo "global_step: ${gs}" && python3 deps/DeepSpeed/deepspeed/checkpoint/ds_to_universal.py --input_folder "${ckpt_dir}" --output_folder ucp
global_step: 95800
[2025-02-07 09:17:47,740] [INFO] [real_accelerator.py:222:get_accelerator] Setting ds_accelerator to xpu (auto detect)
[2025-02-07 09:17:47,968] [INFO] [real_accelerator.py:222:get_accelerator] Setting ds_accelerator to xpu (auto detect)
args = Namespace(input_folder='checkpoints/ws768_ds_stage1_nl32_hs4096_mb4_seq4096_gb3072_sp1_pp1_tp1_bf16_optadamw_lr0.00020_lwf0.05/global_step95800', output_folder='ucp', num_extract_workers=4, num_merge_workers=2, keep_temp_folder=False, strict=True, inject_missing_state=False)
Convert DeepSpeed Checkpoint to Universal Checkpoint
Converting DeepSpeed checkpoint in checkpoints/ws768_ds_stage1_nl32_hs4096_mb4_seq4096_gb3072_sp1_pp1_tp1_bf16_optadamw_lr0.00020_lwf0.05/global_step95800 to Universal checkpoint in ucp
*** 1. Extracting ZeRO fragments
100%|████████████████████| 768/768 [01:48<00:00,  7.10it/s]
*** 2. Merging slices .....
100%|████████████████████| 195/195 [01:03<00:00,  3.06it/s]
*** 3. Saving common optimizer states
*** Done!
took: 0h:03m:41s
```

Once I figure out how to sign the commits appropriately I'll be happy to close this.

Thanks everyone for looking into and fixing this! Super appreciated.

**Comment by loadams** [comment]
> This appears to be fixed, checkpoints which failed to convert previously now seem to be converting successfully.
> 
> e.g.
> 
> ```shell
> #[🐍 aurora_nre_models_frameworks-2024.2.1_u1](👻 aurora_nre_models_frameworks-2024.2.1_u1)
> #[09:17:36 AM][x4609c0s6b0n0][/f/A/f/p/d/Megatron-DeepSpeed][🌱 main][✓] [⏱️ 1m24s]
> $ ckpt_dir=checkpoints/ws768_ds_stage1_nl32_hs4096_mb4_seq4096_gb3072_sp1_pp1_tp1_bf16_optadamw_lr0.00020_lwf0.05/global_step95800 && gs=$(cat "${ckpt_dir}/latest_checkpointed_iteration.txt") && echo "global_step: ${gs}" && python3 deps/DeepSpeed/deepspeed/checkpoint/ds_to_universal.py --input_folder "${ckpt_dir}" --output_folder ucp
> global_step: 95800
> [2025-02-07 09:17:47,740] [INFO] [real_accelerator.py:222:get_accelerator] Setting ds_accelerator to xpu (auto detect)
> [2025-02-07 09:17:47,968] [INFO] [real_accelerator.py:222:get_accelerator] Setting ds_accelerator to xpu (auto detect)
> args = Namespace(input_folder='checkpoints/ws768_ds_stage1_nl32_hs4096_mb4_seq4096_gb3072_sp1_pp1_tp1_bf16_optadamw_lr0.00020_lwf0.05/global_step95800', output_folder='ucp', num_extract_workers=4, num_merge_workers=2, keep_temp_folder=False, strict=True, inject_missing_state=False)
> Convert DeepSpeed Checkpoint to Universal Checkpoint
> Converting DeepSpeed checkpoint in checkpoints/ws768_ds_stage1_nl32_hs4096_mb4_seq4096_gb3072_sp1_pp1_tp1_bf16_optadamw_lr0.00020_lwf0.05/global_step95800 to Universal checkpoint in ucp
> *** 1. Extracting ZeRO fragments
> 100%|████████████████████| 768/768 [01:48<00:00,  7.10it/s]
> *** 2. Merging slices .....
> 100%|████████████████████| 195/195 [01:03<00:00,  3.06it/s]
> *** 3. Saving common optimizer states
> *** Done!
> took: 0h:03m:41s
> ```
> 
> Once I figure out how to sign the commits appropriately I'll be happy to close this.
> 
> Thanks everyone for looking into and fixing this! Super appreciated.

Hi @saforem2 - we can override that if you think this is ready to be merged?

**Comment by saforem2** [comment]
Hey @loadams, I spoke with @tjruwase and @xylian86 last week and I believe this fix was added in [f5735c6](https://github.com/traincheck-team/DeepSpeed/commit/f5735c6ffe9bfab593e25872c29e7007f6547722).

I believe this should now be resolved and I'm happy to close it (pending any objections?)

**Comment by loadams** [comment]
> Hey @loadams, I spoke with @tjruwase and @xylian86 last week and I believe this fix was added in [f5735c6](https://github.com/traincheck-team/DeepSpeed/commit/f5735c6ffe9bfab593e25872c29e7007f6547722).
> 
> I believe this should now be resolved and I'm happy to close it (pending any objections?)

No problem, I hadn't followed the PR discussion, just wanted to help on the DCO changes.  If its already in, feel free to close!
